### PR TITLE
Upgrade to amphp/http-client

### PR DESCRIPTION
**Pull request by enumag**
Trying to upgrade this library from using `amphp/artax` to `amphp/http-client`.

### CODE REVIEW

**Comment by enumag** [comment]
@prolic Ok, it is starting to somewhat work... that is on CI. Locally when I try to run the tests in Docker I'm still getting errors like this:

```
1) ProophTest\EventStoreClient\PersistentSubscriptionManagement\persistent_subscription_manager::can_describe_persistent_subscription
Prooph\EventStore\Exception\ConnectionClosed: Connection '67fcb1577c11c84dabbeb8f67bdc2a2f' was closed

/app/vendor/prooph/event-store/src/Exception/ConnectionClosed.php:20
/app/src/Internal/OperationsManager.php:71
/app/src/Internal/EventStoreConnectionLogicHandler.php:283
/app/src/Internal/EventStoreConnectionLogicHandler.php:526
/app/src/Internal/EventStoreConnectionLogicHandler.php:200
/app/vendor/amphp/amp/lib/Loop/NativeDriver.php:121
/app/vendor/amphp/amp/lib/Loop/Driver.php:134
/app/vendor/amphp/amp/lib/Loop/Driver.php:72
/app/vendor/amphp/amp/lib/Loop.php:84
/app/vendor/amphp/amp/lib/functions.php:182
/app/tests/SpecificationWithConnection.php:54
/app/tests/PersistentSubscriptionManagement/persistent_subscription_manager.php:119
```

Which is of course completely unhelpful. Any idea what to do in these cases?

----

As for fixing the library, the main issue now seems to be that `\Amp\Http\Client\Response::getBody()` returns a `Payload` object instead of a string and the actual body is get through the `Payload::buffer()` method which returns `Promise<string>`. I managed to fix that (see [this commit](https://github.com/prooph/event-store-client/pull/100/commits/b1bbfc9c21f0942105eb611272df061499bda041?w=1)) but I can't say I like the code. Is there any better solution? The same fix needs to be applied in many other places so doing it like this everywhere is not ideal.

**Comment by enumag** [comment]
Alright it seems I managed to fix the tests. At least on CI. Locally I'm still getting the same "ConnectionClosed" issue and I have no idea what's wrong. Any tips appreciated.

**Comment by prolic** [comment]
thanks @enumag  - I try to have a deeper look asap, currently I have visitors from Germany in my house.

File: src/Internal/ClusterDnsEndPointDiscoverer.php [hunk 1]
```diff
@@ -90,8 +90,8 @@ public function __construct(
         $this->gossipTimeout = $gossipTimeout;
         $this->preferRandomNode = $preferRandomNode;
 
-        $this->client = new DefaultClient();
-        $this->client->setOption(Client::OP_TRANSFER_TIMEOUT, $gossipTimeout);
+        $this->client = new Client();
+        $this->client->addApplicationInterceptor(new SetRequestTimeout($gossipTimeout, $gossipTimeout, $gossipTimeout));
```
Comment (enumag) [inline src/Internal/ClusterDnsEndPointDiscoverer.php]: It feels odd to use the timeout variable for 3 different timeouts here but I'm really unsure what else to do here...

**Comment by enumag** [comment]
@prolic Sure, take your time. Not like I need it soon anyway. I'm just trying to get a better understanding of all these async things so this was a nice challenge for me. I'd really appreciate some help with the closed connection issue though.

File: src/Internal/ClusterDnsEndPointDiscoverer.php (hunk 1 above)
Comment (prolic) [reply src/Internal/ClusterDnsEndPointDiscoverer.php]: Yes, just leave it as is, that's fine.

**Comment by prolic** [comment]
Testing locally:

```
sasa@x1 ~/code/event-store-client (git)-[enumag-feature/http-client] % ./vendor/bin/phpunit --exclude-group=ignore --stop-on-failure
PHPUnit 8.3.5 by Sebastian Bergmann and contributors.

...............................................................  63 / 482 ( 13%)
............................................................... 126 / 482 ( 26%)
............................................................... 189 / 482 ( 39%)
............................................................... 252 / 482 ( 52%)
............................................................... 315 / 482 ( 65%)
............................................................... 378 / 482 ( 78%)
............................................................... 441 / 482 ( 91%)
.........................................                       482 / 482 (100%)

Time: 1.9 minutes, Memory: 46.00 MB

OK (482 tests, 1331 assertions)
./vendor/bin/phpunit --exclude-group=ignore --stop-on-failure  31.02s user 12.50s system 37% cpu 1:54.52 total
```

so cannot reproduce your issue

**Comment by enumag** [comment]
Can you try with Docker?

**Comment by prolic** [comment]
Did you follow the readme [here](https://github.com/prooph/event-store-client/tree/master/docker/unittest)? It works for me like a charm.

Maybe you did a `docker run ...` command manually and the default phpunit.xml file has the event-store hosted at `localhost` which is not found of course within docker?

File: src/Internal/ClusterDnsEndPointDiscoverer.php [hunk 2]
```diff
@@ -245,7 +245,7 @@ private function tryGetGossipFrom(GossipSeed $endPoint): Promise
 
                 if (! empty($header)) {
                     $headerData = \explode(':', $header);
-                    $request = $request->withHeader($headerData[0], $headerData[1]);
+                    $request->setHeader($headerData[0], $headerData[1]);
```
Comment (prolic) [inline src/Internal/ClusterDnsEndPointDiscoverer.php]: see https://github.com/amphp/http-client/issues/191

**Comment by enumag** [comment]
> Did you follow the readme here ? It works for me like a charm.
> Maybe you did a docker run ... command manually and the default phpunit.xml file has the event-store hosted at localhost which is not found of course within docker?

No, I followed the guide you linked. I just did it multiple times like try with the artax version, tests passed (well all except two anyway), then composer update, fix a few things, restart docker as noted there and try again. Well ok if it works for you...

**Comment by prolic** [comment]
About the 2 failing test cases: most likely you didn't add `--exclude-group=ignore`, so you run those tests that need db-isolation in the same run.

About tests in docker, see here:

```
sasa@x1 ~/code/event-store-client (git)-[master] % docker-compose -f docker/unittest/docker-compose.yml run php /app/vendor/bin/phpunit -c phpunit/phpunit.xml --exclude-group=ignore
Starting unittest_eventstore_1 ... done
PHPUnit 8.3.5 by Sebastian Bergmann and contributors.

...............................................................  63 / 482 ( 13%)
............................................................... 126 / 482 ( 26%)
............................................................... 189 / 482 ( 39%)
............................................................... 252 / 482 ( 52%)
............................................................... 315 / 482 ( 65%)
............................................................... 378 / 482 ( 78%)
............................................................... 441 / 482 ( 91%)
.........................................                       482 / 482 (100%)

Time: 1.18 minutes, Memory: 48.00 MB

OK (482 tests, 1331 assertions)
sasa@x1 ~/code/event-store-client (git)-[master] % docker-compose -f docker/unittest/docker-compose.yml up -d
130 sasa@x1 ~/code/event-store-client (git)-[master] % docker-compose -f docker/unittest/docker-compose.yml restart                                                                      
Restarting unittest_php_1        ... done
Restarting unittest_eventstore_1 ... done
sasa@x1 ~/code/event-store-client (git)-[master] % docker-compose -f docker/unittest/docker-compose.yml run php /app/vendor/bin/phpunit -c phpunit/phpunit.xml --exclude-group=ignore
Starting unittest_eventstore_1 ... done
PHPUnit 8.3.5 by Sebastian Bergmann and contributors.

...............................................................  63 / 482 ( 13%)
............................................................... 126 / 482 ( 26%)
............................................................... 189 / 482 ( 39%)
............................................................... 252 / 482 ( 52%)
............................................................... 315 / 482 ( 65%)
............................................................... 378 / 482 ( 78%)
............................................................... 441 / 482 ( 91%)
.........................................                       482 / 482 (100%)

Time: 1.17 minutes, Memory: 48.00 MB

OK (482 tests, 1331 assertions)
```

File: composer.json [hunk 3]
```diff
@@ -22,10 +22,10 @@
     "php": "^7.2",
     "ext-json": "*",
     "amphp/amp": "^2.1.2",
-    "amphp/artax": "^3.0.14",
+    "amphp/http-client": "^4.0.0@rc",
```
Comment (enumag) [inline composer.json]:
```suggestion
    "amphp/http-client": "^4.0.0",
```

@prolic Commit this before merging when the stable is out.

File: src/Internal/ClusterDnsEndPointDiscoverer.php [hunk 4]
```diff
@@ -258,7 +258,7 @@ private function tryGetGossipFrom(GossipSeed $endPoint): Promise
                 return null;
             }
 
-            $json = yield $response->getBody()->getInputStream()->read();
+            $json = yield $response->getBody()->read();
```
Comment (enumag) [inline src/Internal/ClusterDnsEndPointDiscoverer.php]: @prolic Can you take a look at this and the `$deferred->resolve($response->getBody()->buffer());` calls bellow? I'm not sure when to use `read` and when to use `buffer`.

File: src/Internal/ClusterDnsEndPointDiscoverer.php (hunk 1 above)
Comment (enumag) [inline src/Internal/ClusterDnsEndPointDiscoverer.php]:
```suggestion
        $this->client = $this->client->withApplicationInterceptor(new SetRequestTimeout($gossipTimeout, $gossipTimeout, $gossipTimeout));
```

File: composer.json (hunk 3 above)
Comment (enumag) [inline composer.json]:
```suggestion
    "amphp/http-client": "^4.0.0@rc6",
```

File: composer.json [hunk 5]
```diff
@@ -22,10 +22,10 @@
     "php": "^7.2",
     "ext-json": "*",
     "amphp/amp": "^2.1.2",
-    "amphp/artax": "^3.0.14",
+    "amphp/http-client": "^4.0.0@rc6",
```
Comment (enumag) [inline composer.json]:
```suggestion
    "amphp/http-client": "^4.0.0-RC6",
```

**Comment by prolic** [comment]
@enumag it now has merge conflict with master, wanna check?

**Comment by enumag** [comment]
Sure. I'll solve it on Monday.

**Comment by enumag** [comment]
@prolic It was easy to solve simply using GitHub UI so I did that. ;-) Luckily `amp/http-client` RC8 and RC9 didn't break anything.

**Comment by enumag** [comment]
@prolic Updated to stable, fixed cs issues

**Comment by enumag** [comment]
ping @prolic

File: src/Internal/ClusterDnsEndPointDiscoverer.php (hunk 4 above)
Comment (prolic) [reply src/Internal/ClusterDnsEndPointDiscoverer.php]: `buffer()` reads everything at once while `read()` only reads a chunk.

**Pull request merged**